File: apps/cli/src/commands/space.ts
import { Command } from "commander";

import { getGitRemoteUrl } from "../lib/detect-space";
import { output, outputQuiet } from "../lib/output";
import { getServerUrl } from "../lib/store";

function requireServer(): string {
    const serverUrl = getServerUrl();
    if (!serverUrl) {
        console.error('No server URL configured. Run "fubbik init" first.');
        process.exit(1);
    }
    return serverUrl;
}

const addSpace = new Command("add")
    .description("Register a space with the server")
    .argument("<name>", "space name")
    .action(async (name: string, _opts: Record<string, unknown>, cmd: Command) => {
        const serverUrl = requireServer();
        const remoteUrl = getGitRemoteUrl();
        const localPath = process.cwd();

        const body: Record<string, unknown> = { name, kind: "code" };
        if (remoteUrl) body.remoteUrl = remoteUrl;
        body.localPaths = [localPath];

        const res = await fetch(`${serverUrl}/api/spaces`, {
            method: "POST",
            headers: { "Content-Type": "application/json" },
            body: JSON.stringify(body)
        });

        if (!res.ok) {
            const text = await res.text();
            console.error(`Failed to create space: ${res.status} ${text}`);
            process.exit(1);
        }

        const data = (await res.json()) as { id: string; name: string };
        outputQuiet(cmd, data.id);
        output(cmd, data, `Created space "${data.name}" (${data.id})`);
    });

const listSpaces = new Command("list")
    .description("List all spaces")
    .action(async (_opts: Record<string, unknown>, cmd: Command) => {
        const serverUrl = requireServer();

        const res = await fetch(`${serverUrl}/api/spaces`);
        if (!res.ok) {
            console.error(`Failed to list spaces: ${res.status}`);
            process.exit(1);
        }

        const data = (await res.json()) as { id: string; name: string; remoteUrl?: string }[];
        outputQuiet(cmd, data.map(c => c.id).join("\n"));

        if (data.length === 0) {
            output(cmd, data, "No spaces found.");
        } else {
            const lines = [`${data.length} space(s):\n`];
            for (const sp of data) {
                const remote = sp.remoteUrl ? ` (${sp.remoteUrl})` : "";
                lines.push(`  ${sp.id}  ${sp.name}${remote}`);
            }
            output(cmd, data, lines.join("\n"));
        }
    });

const removeSpace = new Command("remove")
    .description("Remove a space")
    .argument("<name>", "space name")
    .option("-f, --force", "skip confirmation prompt")
    .action(async (name: string, opts: { force?: boolean }, cmd: Command) => {
        const serverUrl = requireServer();

        // Look up space by name
        const listRes = await fetch(`${serverUrl}/api/spaces`);
        if (!listRes.ok) {
            console.error(`Failed to list spaces: ${listRes.status}`);
            process.exit(1);
        }

        const spaces = (await listRes.json()) as { id: string; name: string }[];
        const match = spaces.find(c => c.name === name);
        if (!match) {
            console.error(`Space "${name}" not found.`);
            process.exit(1);
        }

        if (!opts.force) {
            const readline = await import("node:readline");
            const rl = readline.createInterface({
                input: process.stdin,
                output: process.stdout
            });
            const answer = await new Promise<string>(resolve => {
                rl.question(
                    `This will unlink all chunks from space "${name}". Continue? [y/N] `,
                    resolve
                );
            });
            rl.close();
            if (answer.toLowerCase() !== "y") {
                console.error("Aborted.");
                process.exit(1);
            }
        }

        const delRes = await fetch(`${serverUrl}/api/spaces/${match.id}`, {
            method: "DELETE"
        });
        if (!delRes.ok) {
            console.error(`Failed to delete space: ${delRes.status}`);
            process.exit(1);
        }

        output(cmd, { id: match.id, name }, `Removed space "${name}" (${match.id})`);
    });

const currentSpace = new Command("current")
    .description("Detect the space for the current directory")
    .action(async (_opts: Record<string, unknown>, cmd: Command) => {
        const { detectSpace } = await import("../lib/detect-space");
        const result = await detectSpace();

        if (!result) {
            console.error("No space detected for this directory.");
            process.exit(1);
        }

        outputQuiet(cmd, result.id);
        output(cmd, result, `Current space: "${result.name}" (${result.id})`);
    });

export const spaceCommand = new Command("space")
    .description("Manage spaces")
    .addCommand(addSpace)
    .addCommand(listSpaces)
    .addCommand(removeSpace)
    .addCommand(currentSpace);

File: crates/fubbik-cli/src/commands/space.rs
use std::process::Command;

use anyhow::{Result, bail};
use comfy_table::{Table, presets::UTF8_FULL};
use owo_colors::OwoColorize;

use crate::SpaceCommand;
use crate::client::{Client, Space};
use crate::output::{self, OutputMode};

pub async fn run(client: &Client, command: SpaceCommand, mode: OutputMode) -> Result<()> {
    match command {
        SpaceCommand::List => render_list(client.list_spaces().await?, mode),
        SpaceCommand::Add { name } => {
            let cwd = std::env::current_dir()?;
            let local_path = cwd.to_string_lossy();
            let remote = git_remote();
            let space = client
                .create_space(&name, Some(&local_path), remote.as_deref())
                .await?;
            render_one(&space, "created", mode)
        }
        SpaceCommand::Remove { name, force } => {
            let spaces = client.list_spaces().await?;
            let space = spaces
                .into_iter()
                .find(|space| space.name == name)
                .ok_or_else(|| anyhow::anyhow!("space {name:?} not found"))?;
            if !force {
                bail!("refusing to remove a space without --force");
            }
            let result = client.delete_space(&space.id).await?;
            match mode {
                OutputMode::Json => output::json(&result),
                OutputMode::Quiet => {
                    println!("{}", space.id);
                    Ok(())
                }
                OutputMode::Human => {
                    println!("{} {}", "removed".green(), space.name);
                    Ok(())
                }
            }
        }
        SpaceCommand::Current => {
            let cwd = std::env::current_dir()?;
            let local_path = cwd.to_string_lossy();
            let remote = git_remote();
            let space = client
                .detect_space(Some(&local_path), remote.as_deref())
                .await?
                .ok_or_else(|| anyhow::anyhow!("no space detected for this directory"))?;
            render_one(&space, "current", mode)
        }
    }
}

fn git_remote() -> Option<String> {
    let output = Command::new("git")
        .args(["config", "--get", "remote.origin.url"])
        .output()
        .ok()?;
    output
        .status
        .success()
        .then(|| String::from_utf8_lossy(&output.stdout).trim().to_owned())
        .filter(|value| !value.is_empty())
}

fn render_list(spaces: Vec<Space>, mode: OutputMode) -> Result<()> {
    match mode {
        OutputMode::Json => output::json(&spaces),
        OutputMode::Quiet => {
            for space in spaces {
                println!("{}", space.id);
            }
            Ok(())
        }
        OutputMode::Human => {
            if spaces.is_empty() {
                println!("No spaces found.");
                return Ok(());
            }
            let mut table = Table::new();
            table.load_preset(UTF8_FULL);
            table.set_header(vec!["ID", "Kind", "Name"]);
            for space in spaces {
                table.add_row(vec![space.id, space.kind, space.name]);
            }
            println!("{table}");
            Ok(())
        }
    }
}

fn render_one(space: &Space, verb: &str, mode: OutputMode) -> Result<()> {
    if !output::id_or_json(mode, &space.id, space)? {
        println!("{} {} {}", verb.green(), space.id.dimmed(), space.name);
    }
    Ok(())
}
